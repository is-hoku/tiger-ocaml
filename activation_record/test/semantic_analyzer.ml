open Lexical_analysis
module Lexer_token = Lexical_analysis.Token
open Semantic_action
module Parser_token = Semantic_action.Token
module Types = Semantic_analysis.Types
module Error = Semantic_analysis.Error
open Activation_record
module Env' = Activation_record.Env

let tokenize buf = Parser_token.parsertoken_of_tokent (Lexer.read buf)
let parse buf = Parser.main tokenize buf

let semant venv tenv inside_loop exp level =
  Semant.trans_exp (venv, tenv, inside_loop, exp, level)

let () =
  let fin = open_in "../../example_program/func1.tig" in
  let buf = Lexing.from_channel fin in
  try
    let ast = parse buf in
    Escape.find_escape ast;
    let level = Translate.outermost in
    let expty = semant Env'.base_venv Env'.base_tenv false ast level in
    print_endline (Translate.string_of_level level);
    print_endline (Types.string_of_ty expty.ty);
    close_in fin
  with
  | Lexer.Syntax_error e ->
      print_endline ("SyntaxError: " ^ e);
      close_in fin
  | Parsing.Parse_error ->
      print_endline "SyntaxError";
      close_in fin
  | Error.Type_error -> close_in fin
  | Error.Name_error -> close_in fin
