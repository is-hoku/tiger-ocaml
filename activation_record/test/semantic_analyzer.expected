{index(1), label(102), formals(f0, t102, t103, t104, t105, f5, f6, f7), locals(0)}
{index(3), label(111), formals(f0, t111), locals(0)}
{index(2), label(107), formals(f0, f1, t107, t108, t109, f5), locals(1)}
{index(4), label(113), formals(f0, t113), locals(0)}
{index(0), label(101), formals(), locals(2)}
unit
